### ISSUE TITLE
Questions about the difference between ppo_epochs and epochs in general deep learning training.

### ISSUE THREAD

**Issue by cdxzyc** [lvwerra/trl#129]
In general deep learning training, we generally train `n` epochs and in each epoch we traverse the data in the dataloader  with batch_size `b`.

However, in pro_trainer.py
```
idxs = list(range(bs))
for _ in range(self.config.ppo_epochs):
    random.shuffle(idxs)
    for i in range(bs):
        idx = idxs[i]
        train_stats = self.train_minibatch(
            logprobs[idx].unsqueeze(0),
            values[idx].unsqueeze(0),
            rewards[idx].unsqueeze(0),
            queries[idx].unsqueeze(0),
            responses[idx].unsqueeze(0),
            torch.cat([queries[idx], responses[idx]]).unsqueeze(0),
        )
```
The code calculates the loss for each batch of data  and updates the parameters `ppo_epochs` times successively. How is this different from setting up `n` epoch loops outside the dataloader, and do we need to add `n` epoch loops outside again?

**Comment by lvwerra** [comment]
Closing this for now - feel free to reopen if there are more questions!
